design/one-pager-easy-stacks.md

```diff
@@ -0,0 +1,294 @@
+  signal the readiness of the whole environment.
+
+## Goals
+
+The main goal of this design is to make it easier for people to write sets of
+configurations and be able to manage those configuration sets on a higher level.
+
+It is important that the design puts forward:
+
+* A base boilerplate package where creating a new configuration set that has a
+  controller is as easy as changing the YAML files.
+* A way to declare pre-requisites that needs to be satisfied by the cluster
+  for the configuration set to work.
+* A way to change a high level configuration by changing 1 value that is then
+  translated to the YAMLs that are already deployed.
+
+## Proposal
+
+Crossplane has the notion of stacks that can be used as extension to Crossplane's
+abilites, to deploy an application and all other things that you can do with
+Kubernetes controllers since main components of stacks are CRDs and controllers
+watching them.
+
+To achieve the mentioned goals, we can publish the YAML files in a stack with a
+controller that has its own CRD with high level parameters. Controller deploys
+and updates the YAMLs that are packaged in the stack image according to user's
+input on the CR instance of that stack that user creates.
+
+As sky is the limit when it comes to stacks, we need to put down some general
+rules to be adhered by initial configuration stacks.
+
+* CRDs of the resources that will be deployed by the stack are declared as
+  dependency so that user knows what CR instances will be deployed by the given
+  configuration stack.
+* It should be possible to have different sets of configurations in one stack
+  and user should be able to choose from them via the CR that is an instance
+  of stack's CRD.
+  * Note that it's highly preferred to have 1 custom resource definition that
+    the stack uses and expose the high level variations of the parameters through
+    `spec` instead of having different custom resource definitions for each class
+    of configuration.
+* A minimal high level set of configurations are exposed to the user through
+  the stack's CR, such as `Provider` reference to refer in the deployed resource
+  or `region`.
+* Controller should signal readiness of all resources in CR's status.
+* All resources should be labelled referring to the CR instance of the stack.
+* In YAMLs, resources should refer to each other using cross-resource references
+  wherever possible.
+* **TBD** Controller should update the resources continuously and treat the given CR as
+  the source of truth even though user manually changes the resources that are
+  deployed by the stack.
+* **TBD** CRD of the configuration stack should be namespaced so that different
+  teams/users can create their own configured environment.
+* Stack type will be `ClusterStack` as most of the resources it deploys are
+  cluster-scoped.
+
+Name of the concept is **TBD**; _EasyStack_, _ConfigurationStack_,
```

Comment:
"Resource Pack" is another candidate, and my personal preference : ). I am also liking the "Configuration Stack" option though (or "Configuration Pack").

Reply:
https://www.slideshare.net/SarahZelechoski/how-up-running-culture-is-devaluing-engineering-work

I watched the above presentation recently and it resonated with me. It cautions against branding tools that partially abstract _really complicated stuff_ (like deploying and managing cloud infrastructure) as making those things "easy" - see the deck for more context. Therefore I'd prefer a name that either puts no particular value on what the thing does like `ResourcePack` (this is a pack of resources), or something that emphasises a little more specifically the value they bring. I think they intend to quickly provide infrastructure operators a basic configuration foundation on which to build, so something like `QuickConfig`, or `BaseConfig` could make sense.

Reply:
Are there any emergent patterns around stack names and the custom resources they install?

There's a few _classes of stack_, e.g. "infrastructure stack", "application stack", "template stack". Then there are specific instances of each class of stack, like `stack-gcp`, `stack-azure`, and `stack-wordpress`. We don't have data yet but I suspect many app stacks will be named `stack-foo` and define exactly one custom resource of `kind: Foo`.

So, perhaps:

* This new class of stack is an "infrastructure config stack".
* Each instances of an infrastructure config stack is named like `stack-aws-config-upbound`, or `stack-aws-config-acme` ("this the Upbound variant of the AWS infrastructure configuration stack").
* They each define one custom resource something like `kind: AWSConfigUpbound`.

🤷‍♂

Reply:
@suskin I think the word _stack_ should be in it somewhere. I am leaning towards `Configuration Stacks`.

@negz I feel like `Config` and `Upbound` make the stack names too long. How about? `kind: AWSMinimal group: aws.configurationstacks.upbound.io`

Reply:
> It cautions against branding . . . as making those things "easy"

I agree with @negz wholeheartedly on this. I have a strong dislike of the name "Easy", and the presentation Nic linked explains the reasons better than I was able to articulate them.

> or something that emphasises a little more specifically the value they bring

I see the general name (e.g. Resource Pack, Configuration Stack) as the name for the style of stack, and I see something like the proposed "Base Config" as a more specific term to describe the value that a specific pack/stack is providing. I see them as complementary, and not exclusive. I see it as a corollary to how a `GKECluster` resource is a GKECluster, and is a managed resource, and is a resource.

> I am leaning towards `Configuration Stacks`

I slightly prefer "pack" over "stack" because "stack" just happens to be the format that we're using to bundle it, whereas "pack" describes what it is conceptually, without tying it to a particular format. That makes it more accessible for people who are not yet familiar with our idioms and terms. My **favorite** is "resource pack", but I also like "configuration pack". I slightly prefer "resource" because it is more literal about what is in the pack (a bunch of resources), whereas "configuration" could be interpreted as referring to other types of configuration.